README.md
# reorderable_staggered_scroll_view

[![pub package](https://img.shields.io/pub/v/reorderable_staggered_scroll_view.svg)](https://pub.dartlang.org/packages/reorderable_staggered_scroll_view) [![GitHub license](https://img.shields.io/github/license/itisnajim/reorderable_staggered_scroll_view)](https://github.com/itisnajim/reorderable_staggered_scroll_view/blob/master/LICENSE)  [![GitHub issues](https://img.shields.io/github/issues/itisnajim/reorderable_staggered_scroll_view)](https://github.com/itisnajim/reorderable_staggered_scroll_view/issues)

A Flutter package for creating reorderable and draggable staggered grid and list views. This package provides a smooth and intuitive way to reorder items in both grid and list layouts, making it perfect for creating interactive and user-friendly applications.

| List Demo | Grid Demo |
|:---:|:---:|
| ![List](https://raw.githubusercontent.com/itisnajim/reorderable_staggered_scroll_view/main/readme/list.gif) | ![Grid](https://raw.githubusercontent.com/itisnajim/reorderable_staggered_scroll_view/main/readme/grid.gif) |


## Features

* Reorder items with ease using drag and drop gestures.
* Supports both staggered grid and list layouts.
* Configurable drag-and-drop behavior.
* Designed for a seamless user experience.

## Getting Started

To use this package, add it to your pubspec.yaml file:


```yaml
dependencies:
  reorderable_staggered_scroll_view: ^latest_version
```
Replace ^latest_version with the actual version you want to use.

Import the package in your Dart code:

```dart
import 'package:reorderable_staggered_scroll_view/reorderable_staggered_scroll_view.dart';
```

## Usage

### Grid

```dart
ReorderableStaggeredScrollView.grid(
  enable: _dragEnabled,
  padding: const EdgeInsets.all(16),
  scrollDirection: Axis.vertical,
  physics: const BouncingScrollPhysics(),
  crossAxisCount: 4,
  isLongPressDraggable: false,
  onDragEnd: (details, item) {
    print('onDragEnd: $details ${item.key}');
  },
  children: List.generate(
    10,
    (index) => ReorderableStaggeredScrollViewGridItem(
      key: ValueKey(index.toString()),
      mainAxisCellCount: 1,
      crossAxisCellCount: Random().nextInt(2) + 1,
      widget: Card(
        child: Center(child: Text('Item $index')),
      ),
    ),
  ),
);
```

### List

For a list view of items use `list` constructor:
```dart
ReorderableStaggeredScrollView.list(
  ...
)
```

### Custom

You can integrate the `ReorderableStaggeredScrollView` package with your custom list or grid widget by wrapping your custom list or grid widget with the necessary components.

```dart
DragNotification(
  child: ListView(
    scrollDirection: Axis.vertical,
    children: [
      DragContainer(
        scrollDirection: Axis.vertical,
        isLongPressDraggable: true,
        buildItems: (List<Widget> children) {
          return YourCustomListViewWidget(
            // Your custom list or grid widget configuration
            children: children,
          );
        },
        dataList: yourList, // Your List of ReorderableStaggeredScrollViewListItem or ReorderableStaggeredScrollViewGridItem
        items: (ReorderableStaggeredScrollViewListItem element, DraggableWidget draggableWidget) {
          // Map each item in your dataList into a draggable widget
          return Container(
            key: ValueKey(element.key.toString()),
            child: draggableWidget(element.widget),
          );
        }
      ),
    ],
  ),
);
```
Replace the comments with your actual widget configuration and data. This setup will allow you to use the ReorderableStaggeredScrollView package with your custom widget.

For more detailed usage instructions and configuration options, refer to the [example](https://github.com/itisnajim/reorderable_staggered_scroll_view/tree/main/example) directory in this repository.


## Contributing

We welcome contributions! If you have ideas for improvements or encounter issues, please open an issue or submit a pull request.

## Author

itisnajim, itisnajim@gmail.com

## License
This package is licensed under the MIT License. See the LICENSE file for more details.


CHANGELOG.md
## 0.0.1+3

* Fix onMove callback.

## 0.0.1+2

* Fix event callbacks for the factory grid

## 0.0.1

* TODO: Describe initial release.


test/main.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:reorderable_staggered_scroll_view/reorderable_staggered_scroll_view.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Builder(builder: (context) {
        return Scaffold(
          appBar: AppBar(
            title: const Text('Example'),
          ),
          body: Center(
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: <Widget>[
                ElevatedButton(
                  onPressed: () {
                    Navigator.push(
                      context,
                      MaterialPageRoute(
                        builder: (context) => const HomePage(isList: true),
                      ),
                    );
                  },
                  child: const Padding(
                    padding: EdgeInsets.all(16),
                    child: Text("List"),
                  ),
                ),
                const SizedBox(height: 32),
                ElevatedButton(
                  onPressed: () {
                    Navigator.push(
                      context,
                      MaterialPageRoute(
                        builder: (context) => const HomePage(isList: false),
                      ),
                    );
                  },
                  child: const Padding(
                    padding: EdgeInsets.all(16),
                    child: Text("Grid"),
                  ),
                ),
              ],
            ),
          ),
        );
      }),
    );
  }
}

class HomePage extends StatefulWidget {
  final bool isList;
  const HomePage({Key? key, required this.isList}) : super(key: key);

  @override
  State<HomePage> createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  bool _dragEnabled = true;
  @override
  Widget build(BuildContext context) {
    final nonDraggable = ReorderableStaggeredScrollViewGridItem(
      key: ValueKey(10.toString()),
      mainAxisCellCount: 1,
      crossAxisCellCount: Random().nextInt(2) + 1,
      widget: const Card(
          child: Padding(
        padding: EdgeInsets.all(12),
        child: Center(child: Text('Non Draggable')),
      )),
    );
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.isList ? "List Example" : "Grid Example"),
        actions: [
          IconButton(
              onPressed: () {
                setState(() {
                  _dragEnabled = !_dragEnabled;
                });
              },
              icon: Icon(
                _dragEnabled ? Icons.pause : Icons.play_arrow,
              ))
        ],
      ),
      body: widget.isList
          ? ReorderableStaggeredScrollView.list(
              enable: _dragEnabled,
              padding: const EdgeInsets.all(16),
              physics: const BouncingScrollPhysics(),
              scrollDirection: Axis.vertical,
              axis: Axis.vertical,
              shrinkWrap: true,
              isLongPressDraggable: false,
              onDragEnd: (details, item) {
                debugPrint('onDragEnd: $details ${item.key}');
              },
              isNotDragList: [nonDraggable],
              children: List.generate(
                5,
                (index) => ReorderableStaggeredScrollViewListItem(
                  key: ValueKey(index.toString()),
                  widget: Card(
                      child: Padding(
                    padding: const EdgeInsets.all(12),
                    child: Center(child: Text('Item $index')),
                  )),
                ),
              )..addAll(
                  [
                    nonDraggable,
                    ...List.generate(
                      5,
                      (index) => ReorderableStaggeredScrollViewListItem(
                        key: ValueKey('${index + 5}'),
                        widget: Card(
                            child: Padding(
                          padding: const EdgeInsets.all(12),
                          child: Center(child: Text('Item ${index + 5}')),
                        )),
                      ),
                    )
                  ],
                ),
            )
          : ReorderableStaggeredScrollView.grid(
              enable: _dragEnabled,
              padding: const EdgeInsets.all(16),
              scrollDirection: Axis.vertical,
              physics: const BouncingScrollPhysics(),
              crossAxisCount: 4,
              isLongPressDraggable: false,
              onDragEnd: (details, item) {
                print('onDragEnd: $details ${item.key}');
              },
              isNotDragList: [nonDraggable],
              children: List.generate(
                5,
                (index) => ReorderableStaggeredScrollViewGridItem(
                  key: ValueKey(index.toString()),
                  mainAxisCellCount: 1,
                  crossAxisCellCount: Random().nextInt(2) + 1,
                  widget: Card(child: Center(child: Text('Item $index'))),
                ),
              )..addAll([
                  nonDraggable,
                  ...List.generate(
                    5,
                    (index) => ReorderableStaggeredScrollViewGridItem(
                      key: ValueKey('${index + 5}'),
                      mainAxisCellCount: 1,
                      crossAxisCellCount: Random().nextInt(2) + 1,
                      widget:
                          Card(child: Center(child: Text('Item ${index + 5}'))),
                    ),
                  )
                ]), // Provide the list of reorderable items
            ),
    );
  }
}


test/reorderable_staggered_scroll_view_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'main.dart';

void main() {
  group('ReorderableStaggeredScrollView Tests', () {
    testWidgets('Test if toggling drag enables works',
        (WidgetTester tester) async {
      // Build our app and trigger a frame.
      await tester.pumpWidget(const MaterialApp(home: MyApp()));
      //print('MyApp::');
      // Tap the "Grid" button to navigate to the next page.
      await tester.tap(find.text('Grid'));
      await tester.pumpAndSettle(); // Wait for the navigation to complete.
      //print('Grid::');

      // Find and tap the icon on the next page.
      await tester.tap(find.byIcon(Icons.pause));
      await tester.pumpAndSettle(); // Wait for the icon interaction.
      //print('pause::');

      // Verify that the drag was toggled off.
      expect(find.byIcon(Icons.play_arrow), findsOneWidget);
      //print('play_arrow::');

      // Tap it again to toggle back on.
      await tester.tap(find.byIcon(Icons.play_arrow));
      await tester.pump();
      //print('play_arrow::');

      // Verify that the drag was toggled back on.
      expect(find.byIcon(Icons.pause), findsOneWidget);
      //print('pause::');
    });
  });
}


lib/src/drag_item.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'package:flutter/scheduler.dart';

import 'drag_notification.dart';

/// A widget that animates its child during drag-and-drop operations.
///
/// The [DragItem] widget animates its child using an implicit animation when
/// the child is dragged. It can be used to provide visual feedback during a drag
/// operation.
///
/// - [child]: The child widget that you want to animate during drag-and-drop.
/// - [duration]: The duration of the animation (default is 200 milliseconds).
/// - [onAnimationStatus]: An optional callback to listen for animation status changes.
///
/// To use this widget, wrap it around the widget you want to animate during drag
/// operations, and the animation will be triggered when necessary.
class DragItem extends ImplicitlyAnimatedWidget {
  const DragItem({
    required this.child,
    Duration? duration,
    this.onAnimationStatus,
    Key? key,
  }) : super(key: key, duration: duration ?? _animDuration);
  final Widget child;
  final AnimationStatusListener? onAnimationStatus;
  static const Duration _animDuration = Duration(milliseconds: 200);

  @override
  ImplicitlyAnimatedWidgetState<ImplicitlyAnimatedWidget> createState() =>
      _DragItemState();
}

class _DragItemState extends AnimatedWidgetBaseState<DragItem> {
  RenderAnimManage renderAnimManage = RenderAnimManage();

  @override
  void initState() {
    super.initState();
    renderAnimManage.controller = controller;
    renderAnimManage.animation = animation;
    if (widget.onAnimationStatus != null) {
      controller.addStatusListener(widget.onAnimationStatus!);
    }
  }

  @override
  void forEachTween(TweenVisitor<dynamic> visitor) {}

  void update() {
    controller
      ..value = 0.0
      ..forward();
  }

  @override
  Widget build(BuildContext context) {
    return _DragAnimRender(
      renderAnimManage,
      animation.value,
      change: () {
        SchedulerBinding.instance.addPostFrameCallback((_) {
          update();
        });
      },
      child: widget.child,
    );
  }
}

class _DragAnimRender extends SingleChildRenderObjectWidget {
  const _DragAnimRender(
    this.renderAnimManage,
    this.update, {
    Key? key,
    Widget? child,
    this.change,
  }) : super(key: key, child: child);
  final void Function()? change;
  final RenderAnimManage renderAnimManage;
  final double update;

  bool get isExecute => !DragNotification.isScroll;

  @override
  _AnimRenderObject createRenderObject(BuildContext context) {
    return _AnimRenderObject(renderAnimManage, update, change: change);
  }

  @override
  void updateRenderObject(
      BuildContext context, _AnimRenderObject renderObject) {
    if (isExecute && renderObject.update != update) {
      renderObject.markNeedsLayout();
    }
  }
}

class _AnimRenderObject extends RenderShiftedBox {
  _AnimRenderObject(
    this.renderAnimManage,
    this.update, {
    RenderBox? child,
    this.change,
  }) : super(child);
  final void Function()? change;
  final RenderAnimManage renderAnimManage;
  final double update;

  Offset? lastOffset;

  bool get isExecute => !DragNotification.isScroll;

  @override
  Size computeDryLayout(BoxConstraints constraints) {
    if (child == null) {
      return constraints.constrain(const Size(0, 0));
    }
    final BoxConstraints innerConstraints =
        constraints.deflate(EdgeInsets.zero);
    final Size childSize = child!.getDryLayout(innerConstraints);
    return constraints.constrain(Size(childSize.width, childSize.height));
  }

  @override
  void performLayout() {
    final BoxConstraints constraints = this.constraints;
    if (child == null) {
      size = constraints.constrain(const Size(0, 0));
      return;
    }
    final BoxConstraints innerConstraints =
        constraints.deflate(EdgeInsets.zero);
    child!.layout(innerConstraints, parentUsesSize: true);
    final BoxParentData childParentData = child!.parentData! as BoxParentData;
    childParentData.offset = const Offset(0, 0);
    size = constraints.constrain(Size(
      child!.size.width,
      child!.size.height,
    ));
  }

  void setStart(EdgeInsetsGeometry? begin, EdgeInsetsGeometry? end) {
    renderAnimManage.tweenOffset =
        EdgeInsetsGeometryTween(begin: begin, end: end);
    change?.call();
  }

  @override
  void paint(PaintingContext context, Offset offset) {
    if (child != null) {
      final BoxParentData childParentData = child!.parentData! as BoxParentData;
      final Offset position = childParentData.offset + offset;
      renderAnimManage.currentOffset ??=
          EdgeInsets.only(left: position.dx, top: position.dy);
      if (renderAnimManage.controller.isAnimating &&
          renderAnimManage.tweenOffset != null) {
        final EdgeInsets geometry = renderAnimManage.tweenOffset!
            .evaluate(renderAnimManage.animation) as EdgeInsets;
        context.paintChild(child!, Offset(geometry.left, geometry.top));
        if (renderAnimManage.currentOffset!.left != position.dx ||
            renderAnimManage.currentOffset!.top != position.dy) {
          setStart(
              geometry, EdgeInsets.only(left: position.dx, top: position.dy));
        }
        renderAnimManage.currentOffset =
            EdgeInsets.only(left: position.dx, top: position.dy);
      } else {
        if (isExecute &&
            renderAnimManage.currentOffset != null &&
            (renderAnimManage.currentOffset!.left != position.dx ||
                renderAnimManage.currentOffset!.top != position.dy)) {
          context.paintChild(
              child!,
              Offset(renderAnimManage.currentOffset!.left,
                  renderAnimManage.currentOffset!.top));
          setStart(renderAnimManage.currentOffset,
              EdgeInsets.only(left: position.dx, top: position.dy));
          renderAnimManage.currentOffset =
              EdgeInsets.only(left: position.dx, top: position.dy);
        } else {
          renderAnimManage.currentOffset =
              EdgeInsets.only(left: position.dx, top: position.dy);

          context.paintChild(child!, position);
        }
      }
    }
  }
}

/// A class that manages animation and state for [DragItem].
class RenderAnimManage {
  RenderAnimManage();
  EdgeInsetsGeometryTween? tweenOffset;
  EdgeInsets? currentOffset;
  late AnimationController controller;
  late Animation<double> animation;
}


lib/src/render_box_size.dart
import 'package:flutter/material.dart';

class RenderBoxSize extends StatefulWidget {
  const RenderBoxSize(this.child, this.onChangeSize, {Key? key})
      : super(key: key);

  final Widget child;
  final void Function(Size size) onChangeSize;

  @override
  State<StatefulWidget> createState() => RenderBoxSizeState();
}

class RenderBoxSizeState extends State<RenderBoxSize> {
  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance.addPostFrameCallback((Duration timeStamp) {
      onChangeSize();
    });
  }

  void onChangeSize() {
    final RenderBox? renderBox = context.findRenderObject() as RenderBox?;
    if (renderBox != null) {
      widget.onChangeSize(renderBox.size);
    }
  }

  @override
  Widget build(BuildContext context) {
    return NotificationListener<SizeChangedLayoutNotification>(
        onNotification: (SizeChangedLayoutNotification notification) {
          onChangeSize();
          return true;
        },
        child: widget.child);
  }
}


lib/src/drag_notification.dart
import 'dart:async';

import 'package:flutter/material.dart';

/// A widget that listens to scroll notifications and provides a static property
/// to check whether a scroll action is in progress.
///
/// This widget is designed to be used as a wrapper around another widget to
/// detect scroll actions and set a static `isScroll` property to `true` during
/// scrolling and then back to `false` after a delay.
class DragNotification extends StatefulWidget {
  /// Creates a [DragNotification] widget.
  ///
  /// The [child] is the widget that you want to wrap with scroll detection.
  const DragNotification({required this.child, Key? key}) : super(key: key);

  /// The child widget that will be wrapped by the [DragNotification].
  final Widget child;

  /// A boolean property indicating whether a scroll action is in progress.
  ///
  /// This property is `true` during scrolling and will automatically revert to
  /// `false` after a short delay once the scroll action ends.
  static bool isScroll = false;

  @override
  State<DragNotification> createState() => _DragNotificationState();
}

class _DragNotificationState extends State<DragNotification> {
  Timer? _timer;

  /// Sets the `isScroll` property to `true` and starts a timer to reset it to
  /// `false` after a short delay.
  void setScroll() {
    _timer?.cancel();
    _timer = Timer(const Duration(milliseconds: 100), () {
      DragNotification.isScroll = false;
    });
  }

  @override
  Widget build(BuildContext context) {
    return NotificationListener<ScrollNotification>(
      onNotification: (ScrollNotification notification) {
        if (notification is ScrollStartNotification) {
          _timer?.cancel();
          DragNotification.isScroll = true;
        } else if (notification is ScrollUpdateNotification) {
          _timer?.cancel();
          DragNotification.isScroll = true;
        } else if (notification is ScrollEndNotification) {
          setScroll();
        }
        return false;
      },
      child: widget.child,
    );
  }

  @override
  void dispose() {
    _timer?.cancel();
    super.dispose();
  }
}


example/lib/main.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:reorderable_staggered_scroll_view/reorderable_staggered_scroll_view.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Builder(builder: (context) {
        return Scaffold(
          appBar: AppBar(
            title: const Text('Example'),
          ),
          body: Center(
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: <Widget>[
                ElevatedButton(
                  onPressed: () {
                    Navigator.push(
                      context,
                      MaterialPageRoute(
                        builder: (context) => const HomePage(isList: true),
                      ),
                    );
                  },
                  child: const Padding(
                    padding: EdgeInsets.all(16),
                    child: Text("List"),
                  ),
                ),
                const SizedBox(height: 32),
                ElevatedButton(
                  onPressed: () {
                    Navigator.push(
                      context,
                      MaterialPageRoute(
                        builder: (context) => const HomePage(isList: false),
                      ),
                    );
                  },
                  child: const Padding(
                    padding: EdgeInsets.all(16),
                    child: Text("Grid"),
                  ),
                ),
              ],
            ),
          ),
        );
      }),
    );
  }
}

class HomePage extends StatefulWidget {
  final bool isList;
  const HomePage({Key? key, required this.isList}) : super(key: key);

  @override
  State<HomePage> createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  bool _dragEnabled = true;
  @override
  Widget build(BuildContext context) {
    final nonDraggable = ReorderableStaggeredScrollViewGridItem(
      key: ValueKey(10.toString()),
      mainAxisCellCount: 1,
      crossAxisCellCount: Random().nextInt(2) + 1,
      widget: const Card(
          child: Padding(
        padding: EdgeInsets.all(12),
        child: Center(child: Text('Non Draggable')),
      )),
    );
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.isList ? "List Example" : "Grid Example"),
        actions: [
          IconButton(
              onPressed: () {
                setState(() {
                  _dragEnabled = !_dragEnabled;
                });
              },
              icon: Icon(
                _dragEnabled ? Icons.pause : Icons.play_arrow,
              ))
        ],
      ),
      body: widget.isList
          ? ReorderableStaggeredScrollView.list(
              enable: _dragEnabled,
              padding: const EdgeInsets.all(16),
              physics: const BouncingScrollPhysics(),
              scrollDirection: Axis.vertical,
              axis: Axis.vertical,
              shrinkWrap: true,
              isLongPressDraggable: false,
              onDragEnd: (details, item) {
                debugPrint('onDragEnd: $details ${item.key}');
              },
              isNotDragList: [nonDraggable],
              children: List.generate(
                5,
                (index) => ReorderableStaggeredScrollViewListItem(
                  key: ValueKey(index.toString()),
                  widget: Card(
                      child: Padding(
                    padding: const EdgeInsets.all(12),
                    child: Center(child: Text('Item $index')),
                  )),
                ),
              )..addAll(
                  [
                    nonDraggable,
                    ...List.generate(
                      5,
                      (index) => ReorderableStaggeredScrollViewListItem(
                        key: ValueKey('${index + 5}'),
                        widget: Card(
                            child: Padding(
                          padding: const EdgeInsets.all(12),
                          child: Center(child: Text('Item ${index + 5}')),
                        )),
                      ),
                    )
                  ],
                ),
            )
          : ReorderableStaggeredScrollView.grid(
              enable: _dragEnabled,
              padding: const EdgeInsets.all(16),
              scrollDirection: Axis.vertical,
              physics: const BouncingScrollPhysics(),
              crossAxisCount: 4,
              isLongPressDraggable: false,
              onAccept: (item1, item2, value) {
                print('item1 $item1 item2 $item2 value $value');
              },
              onDragEnd: (details, item) {
                print('onDragEnd: $details ${item.key}');
              },
              onMove: (item, item2, value) {
                print('onMove: item $item item2 $item2 value $value');
              },
              onDragUpdate: (details, item) {
                print('onDragUpdate: details $details item $item');
              },
              isNotDragList: [nonDraggable],
              children: List.generate(
                5,
                (index) => ReorderableStaggeredScrollViewGridItem(
                  key: ValueKey(index.toString()),
                  mainAxisCellCount: 1,
                  crossAxisCellCount: Random().nextInt(2) + 1,
                  widget: Card(child: Center(child: Text('Item $index'))),
                ),
              )..addAll([
                  nonDraggable,
                  ...List.generate(
                    5,
                    (index) => ReorderableStaggeredScrollViewGridItem(
                      key: ValueKey('${index + 5}'),
                      mainAxisCellCount: 1,
                      crossAxisCellCount: Random().nextInt(2) + 1,
                      widget:
                          Card(child: Center(child: Text('Item ${index + 5}'))),
                    ),
                  )
                ]), // Provide the list of reorderable items
            ),
    );
  }
}
